lib/mix/lib/mix/tasks/do.ex

```diff
@@ -15,20 +15,63 @@ defmodule Mix.Tasks.Do do
 
       mix do compile --list, deps
 
-  Note however that the majority of Mix tasks are only
-  executed once per invocation. So for example, the following
-  command will only compile once:
+  You can limit where the tasks are run, in the context of
+  umbrella projects, by passing the app names using --app:
+
+      mix do --app app1 --app app2 compile --list, deps
+
+  Note that the majority of Mix tasks are only executed once
+  per invocation. So for example, the following command will
+  only compile once:
 
       mix do compile, some_other_command, compile
 
   When `compile` is executed again, Mix will notice the task
   has already ran, and skip it.
+
+  ## Command line options
+
+    * `--app` - limit running the tasks to the given app. This
+    option may be given multiple times.
   """
 
   @impl true
   def run(args) do
     Mix.Task.reenable("do")
-    Enum.each(gather_commands(args), fn [task | args] -> Mix.Task.run(task, args) end)
+
+    {apps, args} = extract_apps_from_args(args)
+    show_forgotten_apps_warning(apps)
+
+    Enum.each(gather_commands(args), fn [task | args] ->
+      if apps == [] do
+        Mix.Task.run(task, args)
+      else
+        Mix.Task.run_in_apps(task, apps, args)
+      end
+    end)
+  end
+
+  defp show_forgotten_apps_warning([]), do: nil
+
+  defp show_forgotten_apps_warning(apps) do
+    if Mix.Project.umbrella?() do
+      forgotten_apps = apps -- Enum.map(Mix.Dep.Umbrella.cached(), & &1.app)
+
+      for app <- forgotten_apps do
+        Mix.shell().info([:yellow, "warning: could not find application #{inspect(app)}"])
+      end
+    end
+  end
+
+  defp extract_apps_from_args(args) do
```

Comment:
I think we want to parse the args explicitly. In particular, if we have a task like this:

    mix do --app foo my_task --app bar

The `--app foo` should be captured, but not `--app bar`. Recursion should be enough to do it.

Reply:
You can test this with something like: `mix do cmd echo --app bar` and that should not run inside an app.

Reply:
> I think we want to parse the args explicitly. In particular, if we have a task like this:
> 
> ```
> mix do --app foo my_task --app bar
> ```
> 
> The `--app foo` should be captured, but not `--app bar`. Recursion should be enough to do it.

I think we are already doing exactly that. In fact, I copied the approach from the `cmd` task (where we already have the --app option). It relies on `OptionParser.parse_head!`.

I tested it directly in iex, and here is the result (the first element of the result tuple are the captures):

```ex
iex(16)> OptionParser.parse_head!(["--app", "foo", "my_task", "--app", "bar"], switches: [app: :keep])
{[app: "foo"], ["my_task", "--app", "bar"]}

iex(17)> OptionParser.parse_head!(["cmd", "echo", "--app", "bar"], switches: [app: :keep])
{[], ["cmd", "echo", "--app", "bar"]}
```

Reply:
> You can test this with something like: `mix do cmd echo --app bar` and that should not run inside an app.

See the results running in a test project. It is in line with my previous comment (the --app bar is not captured by the `do` task).

```
➜  mix do cmd echo --app bar
==> second_app
--app bar
==> my_app
--app bar
```

Note:  The `that should not run inside an app` is not accurate. The task is run **in every application**, as it currently works in master (where the --app does not exist at all).

Reply:
Ah, I see. This is a smart trick!